File: apps/mobile/modules/idz-ble-peripheral/android/src/main/java/com/identizen/bleperipheral/IdzBlePeripheralModule.kt
package com.identizen.bleperipheral

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattServer
import android.bluetooth.BluetoothGattServerCallback
import android.bluetooth.BluetoothGattService
import android.bluetooth.BluetoothManager
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.os.Build
import android.os.ParcelUuid
import androidx.core.content.ContextCompat
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID

/*
 * PROTOCOL.md §6.3, Android twin of ios/IdzBlePeripheralModule.swift. The phone advertises the
 * service and serves the 16-byte rotating identifier from one read-only characteristic; Chromium
 * connects, reads it, and POSTs it to /discover/ble. The JS contract (function names, state and
 * authorization strings, event payloads) is identical to the Swift module so src/ble/advertiser.ts
 * needs no platform branches.
 */
private val SERVICE_UUID: UUID = UUID.fromString("F1D0E1A2-1D2E-4B0C-9C0D-1D3E2F4A5B6C")
private val CHARACTERISTIC_UUID: UUID = UUID.fromString("F1D0E1A2-1D2E-4B0C-9C0D-1D3E2F4A5B6D")

class BadRotatingIdException(message: String) : CodedException("ERR_BAD_ROTATING_ID", message, null)

class IdzBlePeripheralModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw CodedException("ERR_NO_CONTEXT", "no React context", null)

  private val manager: BluetoothManager?
    get() = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
  private val adapter: BluetoothAdapter?
    get() = manager?.adapter

  @Volatile private var rotatingId: ByteArray = ByteArray(0)
  @Volatile private var wantAdvertising = false
  @Volatile private var advertising = false
  @Volatile private var lastError: String? = null
  private var gattServer: BluetoothGattServer? = null
  private var serviceAdded = false
  private var advertiser: BluetoothLeAdvertiser? = null
  private var receiver: BroadcastReceiver? = null

  override fun definition() = ModuleDefinition {
    Name("IdzBlePeripheral")

    Events("onStateChange", "onRead")

    OnCreate {
      receiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context?, intent: Intent?) {
          val state = intent?.getIntExtra(BluetoothAdapter.EXTRA_STATE, BluetoothAdapter.ERROR) ?: return
          if (state != BluetoothAdapter.STATE_ON) {
            // The stack drops services when the radio goes down; re-add on the way back up.
            serviceAdded = false
            advertising = false
          }
          reconcile()
          emitState()
        }
      }
      ContextCompat.registerReceiver(
        context,
        receiver,
        IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED),
        ContextCompat.RECEIVER_NOT_EXPORTED,
      )
    }

    OnDestroy {
      receiver?.let { runCatching { context.unregisterReceiver(it) } }
      receiver = null
      stopAdvertising()
      runCatching { gattServer?.close() }
      gattServer = null
      serviceAdded = false
    }

    Function("isSupported") {
      val a = adapter
      a != null && context.packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE) &&
        a.isMultipleAdvertisementSupported
    }

    Function("getState") { snapshot() }

    // Start advertising with the current 16-byte rotating id (hex). Idempotent.
    AsyncFunction("start") { rotatingIdHex: String ->
      rotatingId = parseHex(rotatingIdHex)
      wantAdvertising = true
      lastError = null
      reconcile()
      emitState()
    }

    // Swap the rotating id at a window boundary; the characteristic is answered dynamically.
    AsyncFunction("update") { rotatingIdHex: String ->
      rotatingId = parseHex(rotatingIdHex)
    }

    AsyncFunction("stop") {
      wantAdvertising = false
      stopAdvertising()
      emitState()
    }
  }

  // MARK: Reconcile desired vs actual

  @Synchronized
  private fun reconcile() {
    val a = adapter ?: return
    if (a.state != BluetoothAdapter.STATE_ON) return
    if (!hasPermissions()) {
      lastError = "permission"
      return
    }
    try {
      if (gattServer == null) {
        gattServer = manager?.openGattServer(context, gattCallback)
        serviceAdded = false
      }
      val server = gattServer ?: return
      if (!serviceAdded) {
        val characteristic = BluetoothGattCharacteristic(
          CHARACTERISTIC_UUID,
          BluetoothGattCharacteristic.PROPERTY_READ,
          BluetoothGattCharacteristic.PERMISSION_READ,
        )
        val service = BluetoothGattService(SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)
        service.addCharacteristic(characteristic)
        server.clearServices()
        // Advertising starts in onServiceAdded so the service is live before a central connects.
        server.addService(service)
        serviceAdded = true
        return
      }
      if (wantAdvertising && !advertising) startAdvertising()
      else if (!wantAdvertising && advertising) stopAdvertising()
    } catch (e: SecurityException) {
      lastError = "permission: ${e.message}"
    }
  }

  private fun startAdvertising() {
    val a = adapter ?: return
    val adv = a.bluetoothLeAdvertiser ?: run {
      lastError = "advertising: not supported"
      return
    }
    advertiser = adv
    val settings = AdvertiseSettings.Builder()
      .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
      .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_MEDIUM)
      .setConnectable(true)
      .setTimeout(0)
      .build()
    // Service UUID only: the browser filters on it, and the device name is not ours to change.
    val data = AdvertiseData.Builder()
      .addServiceUuid(ParcelUuid(SERVICE_UUID))
      .setIncludeDeviceName(false)
      .setIncludeTxPowerLevel(false)
      .build()
    try {
      adv.startAdvertising(settings, data, advertiseCallback)
    } catch (e: SecurityException) {
      lastError = "permission: ${e.message}"
      emitState()
    }
  }

  private fun stopAdvertising() {
    try {
      advertiser?.stopAdvertising(advertiseCallback)
    } catch (_: SecurityException) {
      // Nothing to do: without the permission we were never advertising.
    }
    advertising = false
  }

  private val advertiseCallback = object : AdvertiseCallback() {
    override fun onStartSuccess(settingsInEffect: AdvertiseSettings?) {
      advertising = true
      lastError = null
      emitState()
    }

    override fun onStartFailure(errorCode: Int) {
      advertising = false
      lastError = "advertising: " + when (errorCode) {
        ADVERTISE_FAILED_ALREADY_STARTED -> "already started"
        ADVERTISE_FAILED_DATA_TOO_LARGE -> "data too large"
        ADVERTISE_FAILED_FEATURE_UNSUPPORTED -> "unsupported"
        ADVERTISE_FAILED_INTERNAL_ERROR -> "internal error"
        ADVERTISE_FAILED_TOO_MANY_ADVERTISERS -> "too many advertisers"
        else -> "error $errorCode"
      }
      emitState()
    }
  }

  private val gattCallback = object : BluetoothGattServerCallback() {
    override fun onServiceAdded(status: Int, service: BluetoothGattService?) {
      if (status != BluetoothGatt.GATT_SUCCESS) {
        lastError = "service: status $status"
        serviceAdded = false
        emitState()
        return
      }
      reconcile()
    }

    override fun onCharacteristicReadRequest(
      device: BluetoothDevice,
      requestId: Int,
      offset: Int,
      characteristic: BluetoothGattCharacteristic,
    ) {
      val server = gattServer ?: return
      try {
        if (characteristic.uuid != CHARACTERISTIC_UUID) {
          server.sendResponse(device, requestId, BluetoothGatt.GATT_FAILURE, 0, null)
          return
        }
        val id = rotatingId
        if (offset > id.size) {
          server.sendResponse(device, requestId, BluetoothGatt.GATT_INVALID_OFFSET, offset, null)
          return
        }
        server.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, id.copyOfRange(offset, id.size))
        sendEvent(
          "onRead",
          mapOf(
            // The MAC is randomised by the central and only meaningful per session, like the iOS identifier.
            "central" to device.address,
            "at" to System.currentTimeMillis().toDouble(),
          ),
        )
      } catch (e: SecurityException) {
        lastError = "permission: ${e.message}"
        emitState()
      }
    }

    override fun onCharacteristicWriteRequest(
      device: BluetoothDevice,
      requestId: Int,
      characteristic: BluetoothGattCharacteristic,
      preparedWrite: Boolean,
      responseNeeded: Boolean,
      offset: Int,
      value: ByteArray?,
    ) {
      // Nothing is writable.
      if (responseNeeded) {
        try {
          gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_WRITE_NOT_PERMITTED, 0, null)
        } catch (_: SecurityException) {
        }
      }
    }
  }

  // MARK: State

  private fun snapshot(): Map<String, Any?> {
    val body = mutableMapOf<String, Any?>(
      "state" to stateName(),
      "authorization" to authorizationName(),
      "advertising" to advertising,
    )
    lastError?.let { body["error"] = it }
    return body
  }

  private fun emitState() {
    sendEvent("onStateChange", snapshot())
  }

  private fun stateName(): String {
    val a = adapter ?: return "unsupported"
    if (!context.packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE)) return "unsupported"
    if (!hasPermissions()) return "unauthorized"
    return when (a.state) {
      BluetoothAdapter.STATE_ON -> "poweredOn"
      BluetoothAdapter.STATE_TURNING_ON, BluetoothAdapter.STATE_TURNING_OFF -> "resetting"
      BluetoothAdapter.STATE_OFF -> "poweredOff"
      else -> "unknown"
    }
  }

  private fun authorizationName(): String = if (hasPermissions()) "allowedAlways" else "denied"

  private fun hasPermissions(): Boolean {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return true
    val granted = { p: String -> ContextCompat.checkSelfPermission(context, p) == PackageManager.PERMISSION_GRANTED }
    return granted(Manifest.permission.BLUETOOTH_ADVERTISE) && granted(Manifest.permission.BLUETOOTH_CONNECT)
  }

  private fun parseHex(hex: String): ByteArray {
    if (hex.length % 2 != 0 || !hex.all { it.isDigit() || it.lowercaseChar() in 'a'..'f' }) {
      throw BadRotatingIdException("rotating id is not hex: $hex")
    }
    val bytes = ByteArray(hex.length / 2) { i -> hex.substring(i * 2, i * 2 + 2).toInt(16).toByte() }
    if (bytes.size != 16) throw BadRotatingIdException("rotating id must be 16 bytes, got ${bytes.size}")
    return bytes
  }
}

File: apps/mobile/modules/idz-ble-peripheral/ios/IdzBlePeripheralModule.swift
import CoreBluetooth
import ExpoModulesCore

// PROTOCOL.md §6.3: the phone advertises this service; the browser (Chromium Web Bluetooth)
// connects and reads the 16-byte rotating identifier from the characteristic, then POSTs it to
// /discover/ble. Only the index can map the identifier back to a device.
private let kServiceUUID = CBUUID(string: "F1D0E1A2-1D2E-4B0C-9C0D-1D3E2F4A5B6C")
private let kCharacteristicUUID = CBUUID(string: "F1D0E1A2-1D2E-4B0C-9C0D-1D3E2F4A5B6D")
private let kLocalName = "Identizen"

/// Owns the CBPeripheralManager. All calls happen on the main queue.
final class BlePeripheralAdvertiser: NSObject, CBPeripheralManagerDelegate {
  private var manager: CBPeripheralManager?
  private var characteristic: CBMutableCharacteristic?
  private var rotatingId = Data()
  private var wantAdvertising = false
  private var serviceAdded = false
  private var lastError: String?

  /// Module hook: (event name, payload).
  var emit: ((String, [String: Any?]) -> Void)?

  // MARK: Commands

  func start(rotatingId id: Data) {
    rotatingId = id
    wantAdvertising = true
    lastError = nil
    if manager == nil {
      manager = CBPeripheralManager(
        delegate: self,
        queue: nil,
        options: [CBPeripheralManagerOptionShowPowerAlertKey: true]
      )
      // Advertising begins in peripheralManagerDidUpdateState once the radio reports poweredOn.
    } else {
      reconcile()
    }
    emitState()
  }

  func update(rotatingId id: Data) {
    // The characteristic is dynamic (created with a nil value), so the next read returns the new id.
    rotatingId = id
  }

  func stop() {
    wantAdvertising = false
    if let m = manager, m.isAdvertising { m.stopAdvertising() }
    emitState()
  }

  func snapshot() -> [String: Any?] {
    var body: [String: Any?] = [
      "state": stateName(manager?.state ?? .unknown),
      "authorization": authorizationName(),
      "advertising": manager?.isAdvertising ?? false,
    ]
    if let e = lastError { body["error"] = e }
    return body
  }

  // MARK: Reconcile desired vs actual

  private func reconcile() {
    guard let m = manager, m.state == .poweredOn else { return }
    if !serviceAdded {
      let c = CBMutableCharacteristic(
        type: kCharacteristicUUID,
        properties: [.read],
        value: nil, // dynamic: answered in didReceiveRead so rotation needs no re-add
        permissions: [.readable]
      )
      characteristic = c
      let service = CBMutableService(type: kServiceUUID, primary: true)
      service.characteristics = [c]
      m.removeAllServices()
      m.add(service) // advertising starts in didAdd
      serviceAdded = true
      return
    }
    if wantAdvertising && !m.isAdvertising {
      m.startAdvertising([
        CBAdvertisementDataServiceUUIDsKey: [kServiceUUID],
        CBAdvertisementDataLocalNameKey: kLocalName,
      ])
    } else if !wantAdvertising && m.isAdvertising {
      m.stopAdvertising()
    }
  }

  // MARK: CBPeripheralManagerDelegate

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    if peripheral.state != .poweredOn {
      // Services are dropped when the radio goes down; re-add on the way back up.
      serviceAdded = false
    }
    reconcile()
    emitState()
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didAdd service: CBService, error: Error?) {
    if let error = error {
      lastError = "service: \(error.localizedDescription)"
      serviceAdded = false
      emitState()
      return
    }
    reconcile()
  }

  func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
    lastError = error.map { "advertising: \($0.localizedDescription)" }
    emitState()
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveRead request: CBATTRequest) {
    guard request.characteristic.uuid == kCharacteristicUUID else {
      peripheral.respond(to: request, withResult: .attributeNotFound)
      return
    }
    guard request.offset <= rotatingId.count else {
      peripheral.respond(to: request, withResult: .invalidOffset)
      return
    }
    request.value = rotatingId.subdata(in: request.offset..<rotatingId.count)
    peripheral.respond(to: request, withResult: .success)
    emit?("onRead", [
      "central": request.central.identifier.uuidString,
      "at": Date().timeIntervalSince1970 * 1000,
    ])
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
    // Nothing is writable; reject the batch.
    if let first = requests.first {
      peripheral.respond(to: first, withResult: .writeNotPermitted)
    }
  }

  // MARK: Helpers

  private func emitState() {
    emit?("onStateChange", snapshot())
  }

  private func stateName(_ s: CBManagerState) -> String {
    switch s {
    case .unknown: return "unknown"
    case .resetting: return "resetting"
    case .unsupported: return "unsupported"
    case .unauthorized: return "unauthorized"
    case .poweredOff: return "poweredOff"
    case .poweredOn: return "poweredOn"
    @unknown default: return "unknown"
    }
  }

  private func authorizationName() -> String {
    switch CBManager.authorization {
    case .notDetermined: return "notDetermined"
    case .restricted: return "restricted"
    case .denied: return "denied"
    case .allowedAlways: return "allowedAlways"
    @unknown default: return "notDetermined"
    }
  }
}

enum IdzBleError: Error, LocalizedError {
  case badHex(String)
  case badLength(Int)

  var errorDescription: String? {
    switch self {
    case .badHex(let s): return "rotating id is not hex: \(s)"
    case .badLength(let n): return "rotating id must be 16 bytes, got \(n)"
    }
  }
}

private func dataFromHex(_ hex: String) throws -> Data {
  var data = Data(capacity: hex.count / 2)
  var index = hex.startIndex
  while index < hex.endIndex {
    guard let next = hex.index(index, offsetBy: 2, limitedBy: hex.endIndex),
          let byte = UInt8(hex[index..<next], radix: 16) else {
      throw IdzBleError.badHex(hex)
    }
    data.append(byte)
    index = next
  }
  guard data.count == 16 else { throw IdzBleError.badLength(data.count) }
  return data
}

public class IdzBlePeripheralModule: Module {
  private let advertiser = BlePeripheralAdvertiser()

  public func definition() -> ModuleDefinition {
    Name("IdzBlePeripheral")

    Events("onStateChange", "onRead")

    OnCreate {
      self.advertiser.emit = { [weak self] name, body in
        self?.sendEvent(name, body)
      }
    }

    Function("isSupported") { () -> Bool in
      true
    }

    Function("getState") { () -> [String: Any?] in
      self.advertiser.snapshot()
    }

    /// Start advertising with the current 16-byte rotating id (hex). Idempotent.
    AsyncFunction("start") { (rotatingIdHex: String) in
      let id = try dataFromHex(rotatingIdHex)
      self.advertiser.start(rotatingId: id)
    }.runOnQueue(.main)

    /// Swap the rotating id at a window boundary without restarting the radio.
    AsyncFunction("update") { (rotatingIdHex: String) in
      let id = try dataFromHex(rotatingIdHex)
      self.advertiser.update(rotatingId: id)
    }.runOnQueue(.main)

    AsyncFunction("stop") {
      self.advertiser.stop()
    }.runOnQueue(.main)
  }
}
